packages/viewer/src/modules/UrlHelper.ts

```diff
@@ -0,0 +1,310 @@
+    const commitId = url.split('/commits/')[1].substring(0, 10)
+    const objUrl = await this.getCommitReferencedObjectUrl({
+      origin: parsed.origin,
+      streamId,
+      commitId
+    })
+    objsUrls.push(objUrl)
+  }
+
+  // object based urls
+  if (url.includes('objects')) objsUrls.push(url)
+
+  // supports urls that include overlay queries
+  // e.g., https://speckle.xyz/streams/a632e7a784/objects/457c45feffa6f954572e5e86fb6d4f25?overlay=cf8dc76247,f5adc1d991b3dceb4b5ad6b50f919a0e
+  if (url.includes('overlay=')) {
+    const searchParams = new URLSearchParams(parsed.search)
+    const resIds = searchParams.get('overlay')?.split(',')
+    if (resIds !== undefined) {
+      for (const resId of resIds) {
+        if (resId.length === 10) {
+          objsUrls.push(
+            await getCommitReferencedObjectUrl(
+              {
+                origin: parsed.origin,
+                streamId,
+                commitId: resId
+              } as CommitReferencedObjectUrl,
+              authToken
+            )
+          )
+        } else {
+          objsUrls.push(`${parsed.origin}/streams/${streamId}/objects/${resId}`)
+        }
+      }
+    }
+  }
+
+  return objsUrls
+}
+
+async function getCommitReferencedObjectUrl(
+  ref: CommitReferencedObjectUrl,
+  authToken?: string
+) {
+  const headers: { 'Content-Type': string; Authorization: string } = {
+    'Content-Type': 'application/json',
+    Authorization: ''
+  }
+  if (authToken) {
+    headers['Authorization'] = `Bearer ${authToken}`
+  }
+  const res = await fetch(`${ref.origin}/graphql`, {
+    method: 'POST',
+    headers,
+    body: JSON.stringify({
+      query: `
+          query Stream($streamId: String!, $commitId: String!) {
+            stream(id: $streamId) {
+              commit(id: $commitId) {
+                referencedObject
+              }
+            }
+          }
+        `,
+      variables: { streamId: ref.streamId, commitId: ref.commitId }
+    })
+  })
+
+  const { data } = await res.json()
+  return `${ref.origin}/streams/${ref.streamId}/objects/${data.stream.commit.referencedObject}`
+}
+
+async function getNewResourceUrls(url: string, authToken?: string): Promise<string[]> {
+  const objsUrls: string[] | PromiseLike<string[]> = []
+  const parsed = new URL(decodeURI(url))
+  const params = parsed.href.match(/[^/]+$/)
+  if (!params) {
+    return Promise.reject('No model or object ids specified')
+  }
+
+  const projectId = parsed.href.split('/projects/')[1].substring(0, 10)
+  const headers: { 'Content-Type': string; Authorization: string } = {
+    'Content-Type': 'application/json',
+    Authorization: ''
+  }
+
+  if (authToken) {
+    headers['Authorization'] = `Bearer ${authToken}`
+  }
+
+  const ref: ReferencedObjectUrl = {
+    origin: parsed.origin,
+    projectId
+  }
+
+  const resources = SpeckleViewer.ViewerRoute.parseUrlParameters(
+    decodeURIComponent(params[0])
+  )
+
+  for (let k = 0; k < resources.length; k++) {
```

Comment:
Instead of awaiting each of those async functions serially, you should add all of the resulting promises to an array and then `await Promise.all()` them, to ensure they're all loaded in parallel